doc/design/concurrent_data_transfer_and_kernel_execution.md

```diff
@@ -0,0 +1,93 @@
+  
+<p align="center"><img src="images/gpu_async.png" height="75%" width = "75%" align="center"/><br/></p>
+  
+## Solution
+One feasible method is to use [staging area](https://en.wikipedia.org/wiki/Staging_(data)). A staging area is an intermediate storage area used for data processing during the [extract, transform and load (ETL)](https://en.wikipedia.org/wiki/Extract,_transform,_load) process. The data staging area sits between the data source(s) and the data target(s), which are often data warehouses, data marts, or other data repositories.   
+The staging area has one fixed-size buffer and two methods, `put` and `get`. `put` and `get` access buffer in a thread-safe way. The operation of `put` puts data into the buffer, and `get` removes a data from the buffer. When the program runs, the staging area should be warmed up. At first, `put` is called to put some data in the staging area. After that, `get` is called when the kernel begins executing, and one data is extracted from the staging area. At the same time, `put` is called and a new data is copied from CPU side to the staging area. Because the operation of data transfer is asynchronous, so the `put` method will return immediately.     
+
+The GPU provided by NVIDIA generally has two engines: a copy engine and a computing engine. To make two engines working at the same time, we need to use the [stream mechanism](https://devblogs.nvidia.com/parallelforall/gpu-pro-tip-cuda-7-streams-simplify-concurrency/) provided by CUDA. As long as data transfer and kernel execution are in different streams and the kernel does not use the data being transferred, the data transfer and kernel execution can be done at the same time on GPU.  
+
+To support the above description, we need to define a new class: `Channel`. Here, we borrow the concept of [`Channel`](https://www.golang-book.com/books/intro/10#section2) in the go language.
+
+```
+template <typename T>
+class Channel {
+ private:
+
+  std::size_t capacity_;
+  std::size_t bytes_limit_;
+  std::size_t current_bytes_;
+  std::mutex mu_;
+  std::condition_variable empty_cond_var_;
+  std::condition_variable full_cond_var_;
+  std::deque<T> channel_;
+
+ public:
+  explicit Channel(std::size_t capacity, std::size_t bytes_limit);
+
+  void Put(T* buffer_element) { ... }
+  void Get(T* buffer_element) { ... }
+  size_t Size() { ... }
+  void Clear()  { ... }
+};
+};
+```
+
+**Take the [recognize_digits_mlp](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/book/test_recognize_digits_mlp.py) as an example to show the model definition after using the staging area mechanism.**
+
+```
+...
+chan_list_config = fluid.channel_list.config(name="chan_list", type=var.CHANNEL_LIST)
+
+with fluid.go(concurrent_program, chan_list_config) as go:
```

Comment:
I though about this for a while, now I believe that `fluid.go` should start a thread to run a *block* in the current ProgramDesc, instead of a new ProgramDesc, because a ProgramDesc is a program, like a .go file, and Go's go statement doesn't run a .go file, but a lambda represented by a block.